phpBB/phpbb/cp/menu.php

```diff
@@ -0,0 +1,311 @@
+			{
+				// Assign the active category to the template
+				$this->template->assign_vars(["{$cp}_menu_active" => $menu[$category]]);
+			}
+		}
+	}
+
+	/**
+	 * Build the active control panel menu items array.
+	 *
+	 * The accessed menu item and its parents are added to the array,
+	 * where the active control panel category is the last value.
+	 *
+	 * @return void
+	 */
+	protected function build_actives()
+	{
+		$route = $this->symfony_request->get('_route');
+		$route = str_replace($this->cp_manager->get_route_pagination(), '', $route);
+
+		if (array_key_exists($route, $this->collection))
+		{
+			$item = $this->collection[$route];
+
+			$this->actives[] = $route;
+
+			while ($item !== null)
+			{
+				$parent = $item['parent'];
+
+				if ($parent && array_key_exists($parent, $this->collection))
+				{
+					$this->actives[] = $parent;
+
+					$item = $this->collection[$parent];
+				}
+				else
+				{
+					$item = null;
+				}
+			}
+		}
+	}
+
+	/**
+	 * Build the control panel menu items.
+	 *
+	 * Index all items per parent and
+	 * check if an item should be inserted before a specific item.
+	 *
+	 * @return void
+	 */
+	protected function build_menu()
+	{
+		foreach ($this->collection as $name => $item)
+		{
+			$before = $item['before'];
+			$parent = $item['parent'];
+
+			if (!empty($this->items[$parent][$before]))
+			{
+				$this->insert_item_before($name, $item);
+			}
+			else
+			{
+				$this->items[$parent][$name] = $item;
+			}
+		}
+	}
+
+	/**
+	 * Build the control panel navigation menu.
+	 *
+	 * Iterate over the items for the specified parent.
+	 * The default parent '', is for the top level categories.
+	 *
+	 * This checks permissions and creates the template variables per item.
+	 *
+	 * @param string	$parent			The menu item parent
+	 * @return array					The menu items belonging the the specified parent
+	 */
+	protected function build_menu_tree($parent = '')
+	{
+		$menu = [];
+
+		foreach ($this->items[$parent] as $name => $item)
+		{
+			// If the authorisation requirements are met
+			if ($this->cp_auth->check_auth(
+				$item['auth'],
+				$this->cp_ids->get_forum_id(),
+				$this->cp_ids->get_topic_id(),
+				$this->cp_ids->get_post_id()
+			))
+			{
+				$route = $item['route'];
+
+				// Is this a category?
+				$s_category = (bool) is_string($route);
+
+				// Set up this item's template variables
+				$variables = $this->get_item_variables($name, $item);
+
+				// Does this item have any children?
+				if (!empty($this->items[$name]))
+				{
+					// Iterate over all the children
+					$variables['ITEMS'] = $this->build_menu_tree($name);
+
+					// If this is a category and the categories pre-defined route is not available
+					if ($s_category && empty($variables['ITEMS'][$route]))
+					{
+						// Get the first child of the category
+						$first = reset($variables['ITEMS']);
+
+						// Set the first child's route as the category's route
+						$variables['U_VIEW'] = $first['U_VIEW'];
```

Comment:
What I got so far.
On the login page (like `http://localhost/cp/user/index`) some of items data is empty, as following (`var_dump($name);
var_dump($variables['ITEMS']);` output):
```php
string(18) "mcp_cat_moderation" array(0) { } array(0) { }
string(15) "mcp_cat_reports" array(0) { } array(0) { }
string(16) "mcp_cat_warnings" array(0) { } array(0) { }
string(12) "mcp_cat_logs" array(0) { } array(0) { }
string(12) "mcp_cat_bans" array(0) { } array(0) { }
```
Hence `$variables['ITEMS']` is empty and `$first = reset($variables['ITEMS']);` returns false. So actually at this line the code tries to access array offset on bool value.

Reply:
That's a good catch. The problem is not your comment below, as the module installation is no longer used at all. All modules are defined in symfony parameters. So they are present.

The issue was, that the MCP menu was even being build for non-logged in users. So all items in each category did not pass authorisation, as the user is not logged in. So I've added a check to only build the MCP menu if the user is logged in and not a bot: `if ($user->data['is_registered'] && !$user->data['is_bot'])`.

Moreover, I also added a check in this specific block to see if `$variable['ITEMS']` is not empty, to prevent undefined index errors.